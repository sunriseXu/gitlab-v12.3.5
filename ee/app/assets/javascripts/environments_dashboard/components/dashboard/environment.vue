<script>
import _ from 'underscore';
import { GlTooltipDirective, GlLink } from '@gitlab/ui';
import { s__, __, sprintf } from '~/locale';
import Icon from '~/vue_shared/components/icon.vue';
import timeagoMixin from '~/vue_shared/mixins/timeago';
import UserAvatarLink from '~/vue_shared/components/user_avatar/user_avatar_link.vue';
import Commit from '~/vue_shared/components/commit.vue';
import Alerts from 'ee/vue_shared/dashboards/components/alerts.vue';
import TimeAgo from 'ee/vue_shared/dashboards/components/time_ago.vue';
import { STATUS_FAILED, STATUS_RUNNING } from 'ee/vue_shared/dashboards/constants';
import ProjectPipeline from 'ee/vue_shared/dashboards/components/project_pipeline.vue';
import EnvironmentHeader from './environment_header.vue';

export default {
  components: {
    EnvironmentHeader,
    UserAvatarLink,
    GlLink,
    Commit,
    Alerts,
    ProjectPipeline,
    TimeAgo,
    Icon,
  },
  directives: {
    'gl-tooltip': GlTooltipDirective,
  },
  mixins: [timeagoMixin],
  props: {
    environment: {
      type: Object,
      required: true,
    },
  },
  tooltips: {
    timeAgo: __('Finished'),
    triggerer: __('Triggerer'),
    job: s__('EnvironmentsDashboard|Job: %{job}'),
  },
  noDeploymentMessage: __('This environment has no deployments yet.'),
  computed: {
    hasPipelineFailed() {
      return (
        this.lastPipeline &&
        this.lastPipeline.details &&
        this.lastPipeline.details.status &&
        this.lastPipeline.details.status.group === STATUS_FAILED
      );
    },
    hasPipelineErrors() {
      return this.environment.alert_count > 0;
    },
    cardClasses() {
      return {
        'dashboard-card-body-warning': !this.hasPipelineFailed && this.hasPipelineErrors,
        'dashboard-card-body-failed': this.hasPipelineFailed,
        'bg-secondary': !this.hasPipelineFailed && !this.hasPipelineErrors,
      };
    },
    user() {
      return this.lastDeployment && !_.isEmpty(this.lastDeployment.user)
        ? this.lastDeployment.user
        : null;
    },
    lastPipeline() {
      return !_.isEmpty(this.environment.last_pipeline) ? this.environment.last_pipeline : null;
    },
    lastDeployment() {
      return !_.isEmpty(this.environment.last_deployment) ? this.environment.last_deployment : null;
    },
    deployable() {
      return !_.isEmpty(this.lastDeployment.deployable) ? this.lastDeployment.deployable : null;
    },
    commit() {
      return !_.isEmpty(this.lastDeployment.commit) ? this.lastDeployment.commit : {};
    },
    jobTooltip() {
      return sprintf(this.$options.tooltips.job, { job: this.buildName });
    },
    commitRef() {
      return this.lastDeployment && !_.isEmpty(this.lastDeployment.commit)
        ? {
            ...this.lastDeployment.commit,
            ...this.lastDeployment.ref,
            ref_url: this.lastDeployment.ref.ref_path,
          }
        : {};
    },
    finishedTime() {
      return this.deployable.updated_at;
    },
    finishedTimeTitle() {
      return this.tooltipTitle(this.finishedTime);
    },
    shouldShowTimeAgo() {
      return (
        this.deployable &&
        this.deployable.status &&
        this.deployable.status.group !== STATUS_RUNNING &&
        this.finishedTime
      );
    },
    buildName() {
      if (
        this.environment &&
        this.environment.last_deployment &&
        this.environment.last_deployment.deployable
      ) {
        const { deployable } = this.environment.last_deployment;
        return `${deployable.name} #${deployable.id}`;
      }
      return '';
    },
  },
};
</script>
<template>
  <div class="dashboard-card card border-0">
    <environment-header
      :environment="environment"
      :has-pipeline-failed="hasPipelineFailed"
      :has-errors="hasPipelineErrors"
    />

    <div :class="cardClasses" class="dashboard-card-body card-body">
      <div v-if="lastDeployment" class="row">
        <div class="col-1 align-self-center">
          <user-avatar-link
            v-if="user"
            :link-href="user.path"
            :img-src="user.avatar_url"
            :tooltip-text="user.name"
            :img-size="32"
          />
        </div>

        <div class="col-10 col-sm-6 pr-0 pl-5 align-self-center align-middle ci-table">
          <div class="branch-commit">
            <icon name="work" />
            <gl-link v-gl-tooltip="jobTooltip" :href="deployable.build_path" class="str-truncated">
              {{ buildName }}
            </gl-link>
          </div>
          <commit
            :tag="commitRef.tag"
            :commit-ref="commitRef"
            :short-sha="commit.short_id"
            :commit-url="commit.commit_url"
            :title="commit.title"
            :author="commit.author"
            :show-branch="true"
          />
        </div>

        <div class="col-sm-5 pl-0 text-right align-self-center d-none d-sm-block">
          <time-ago
            v-if="shouldShowTimeAgo"
            :time="finishedTime"
            :tooltip-text="$options.tooltips.timeAgo"
          />
          <alerts :count="environment.alert_count" :last-alert="environment.last_alert" />
        </div>

        <div v-if="lastPipeline" class="col-12">
          <project-pipeline
            :project-name="environment.name_with_namespace"
            :last-pipeline="lastPipeline"
            :has-pipeline-failed="hasPipelineFailed"
          />
        </div>
      </div>

      <div v-else class="h-100 d-flex justify-content-center align-items-center">
        <div class="text-plain text-metric text-center bold w-75">
          {{ $options.noDeploymentMessage }}
        </div>
      </div>
    </div>
  </div>
</template>
