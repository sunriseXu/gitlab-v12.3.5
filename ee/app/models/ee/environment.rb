# frozen_string_literal: true

module EE
  module Environment
    extend ActiveSupport::Concern
    extend ::Gitlab::Utils::Override
    include ::Gitlab::Utils::StrongMemoize

    prepended do
      has_many :prometheus_alerts, inverse_of: :environment
      has_one :last_deployable, through: :last_deployment, source: 'deployable', source_type: 'CommitStatus'
      has_one :last_pipeline, through: :last_deployable, source: 'pipeline'

      # Returns environments where its latest deployment is to a cluster
      scope :deployed_to_cluster, -> (cluster) do
        environments = model.arel_table
        deployments = Deployment.arel_table
        later_deployments = Deployment.arel_table.alias('latest_deployments')
        join_conditions = later_deployments[:environment_id]
          .eq(deployments[:environment_id])
          .and(deployments[:id].lt(later_deployments[:id]))

        join = deployments
          .join(later_deployments, Arel::Nodes::OuterJoin)
          .on(join_conditions)

        model
          .joins(:deployments)
          .joins(join.join_sources)
          .where(later_deployments[:id].eq(nil))
          .where(deployments[:cluster_id].eq(cluster.id))
          .where(deployments[:project_id].eq(environments[:project_id]))
      end

      scope :preload_for_cluster_environment_entity, -> do
        preload(
          last_deployment: [:deployable],
          project: [:route, { namespace: :route }]
        )
      end
    end

    def reactive_cache_updated
      super

      ::Gitlab::EtagCaching::Store.new.tap do |store|
        store.touch(
          ::Gitlab::Routing.url_helpers.project_environments_path(project, format: :json))

        store.touch(cluster_environments_etag_key) if cluster_environments_etag_key
      end
    end

    def cluster_environments_etag_key
      strong_memoize(:cluster_environments_key) do
        cluster = last_deployment&.cluster

        if cluster&.group_type?
          ::Gitlab::Routing.url_helpers.environments_group_cluster_path(cluster.group, cluster)
        end
      end
    end

    def pod_names
      return [] unless rollout_status

      rollout_status.instances.map do |instance|
        instance[:pod_name]
      end
    end

    def clear_prometheus_reactive_cache!(query_name)
      cluster_prometheus_adapter&.clear_prometheus_reactive_cache!(query_name, self)
    end

    def cluster_prometheus_adapter
      @cluster_prometheus_adapter ||= Prometheus::AdapterService.new(project, deployment_platform).cluster_prometheus_adapter
    end

    def protected?
      project.protected_environment_by_name(name).present?
    end

    def protected_deployable_by_user?(user)
      project.protected_environment_accessible_to?(name, user)
    end

    def rollout_status
      return unless has_terminals?

      result = with_reactive_cache do |data|
        deployment_platform.rollout_status(self, data)
      end

      result || ::Gitlab::Kubernetes::RolloutStatus.loading
    end
  end
end
