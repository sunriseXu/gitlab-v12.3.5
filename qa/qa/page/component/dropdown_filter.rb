# frozen_string_literal: true

module QA
  module Page
    module Component
      module DropdownFilter
        def filter_and_select(item)
          wait(reload: false) do
            page.has_css?('.dropdown-input-field')
          end

          find('.dropdown-input-field').set(item)
          click_link item
        end
      end
    end
  end
end
