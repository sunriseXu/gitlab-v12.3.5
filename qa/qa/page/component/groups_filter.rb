# frozen_string_literal: true

module QA
  module Page
    module Component
      module GroupsFilter
        def self.included(base)
          base.view 'app/views/shared/groups/_search_form.html.haml' do
            element :groups_filter
          end

          base.view 'app/assets/javascripts/groups/components/groups.vue' do
            element :groups_list_tree_container
          end
        end

        private

        def has_filtered_group?(name)
          # Filter and submit to reload the page and only retrieve the filtered results
          find_element(:groups_filter).set(name).send_keys(:return)

          # Since we submitted after filtering, the presence of
          # groups_list_tree_container means we have the complete filtered list
          # of groups
          wait(reload: false) do
            page.has_css?(element_selector_css(:groups_list_tree_container))
          end

          # If there are no groups we'll know immediately because we filtered the list
          return false if page.has_text?('No groups or projects matched your search', wait: 0)

          # The name will be present as filter input so we check for a link, not text
          page.has_link?(name, wait: 0)
        end
      end
    end
  end
end
