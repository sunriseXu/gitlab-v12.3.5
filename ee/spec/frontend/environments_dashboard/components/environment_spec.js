import { shallowMount, createLocalVue } from '@vue/test-utils';
import Commit from '~/vue_shared/components/commit.vue';
import component from 'ee/environments_dashboard/components/dashboard/environment.vue';
import EnvironmentHeader from 'ee/environments_dashboard/components/dashboard/environment_header.vue';
import Alert from 'ee/vue_shared/dashboards/components/alerts.vue';
import environment from './mock_environment.json';

const localVue = createLocalVue();

describe('Environment', () => {
  const Component = localVue.extend(component);
  let wrapper;
  let propsData;

  beforeEach(() => {
    propsData = {
      environment,
    };
    wrapper = shallowMount(Component, {
      localVue,
      propsData,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('matchs the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('wrapped components', () => {
    describe('environment header', () => {
      it('binds environment', () => {
        expect(wrapper.find(EnvironmentHeader).props('environment')).toBe(environment);
      });
    });
    describe('alerts', () => {
      let alert;

      beforeEach(() => {
        alert = wrapper.find(Alert);
      });
      it('binds alert count to count', () => {
        expect(alert.props('count')).toBe(environment.alert_count);
      });
      it('binds last alert', () => {
        expect(alert.props('lastAlert')).toBe(environment.last_alert);
      });
    });
    describe('commit', () => {
      let commit;

      beforeEach(() => {
        commit = wrapper.find(Commit);
      });

      it('binds commitRef', () => {
        expect(commit.props('commitRef')).toBe(wrapper.vm.commitRef);
      });

      it('binds short_id to shortSha', () => {
        expect(commit.props('shortSha')).toBe(environment.last_deployment.commit.short_id);
      });

      it('binds commitUrl', () => {
        expect(commit.props('commitUrl')).toBe(environment.last_deployment.commit.commit_url);
      });

      it('binds title', () => {
        expect(commit.props('title')).toBe(environment.last_deployment.commit.title);
      });

      it('binds author', () => {
        expect(commit.props('author')).toBe(environment.last_deployment.commit.author);
      });

      it('binds tag', () => {
        expect(commit.props('tag')).toBe(environment.last_deployment.ref.tag);
      });
    });
  });
});
