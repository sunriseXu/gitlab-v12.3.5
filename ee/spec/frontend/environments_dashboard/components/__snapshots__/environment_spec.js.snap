// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Environment matchs the snapshot 1`] = `
<div
  class="dashboard-card card border-0"
>
  <environmentheader-stub
    environment="[object Object]"
    haserrors="true"
  />
   
  <div
    class="dashboard-card-body card-body dashboard-card-body-warning"
  >
    <div
      class="row"
    >
      <div
        class="col-1 align-self-center"
      >
        <useravatarlink-stub
          imgalt=""
          imgcssclasses=""
          imgsize="32"
          imgsrc="https://secure.gravatar.com/avatar/e64c7d89f26bd1972efa854d13d7dd61?s=80&d=identicon"
          linkhref="/root"
          tooltipplacement="top"
          tooltiptext="Administrator"
          username=""
        />
      </div>
       
      <div
        class="col-10 col-sm-6 pr-0 pl-5 align-self-center align-middle ci-table"
      >
        <div
          class="branch-commit"
        >
          <icon-stub
            cssclasses=""
            name="work"
            size="16"
          />
           
          <gllink-stub
            class="str-truncated"
            data-original-title=""
            href="/root/minimal-ruby-app/-/jobs/1097"
            title=""
          >
            
            production #1097
          
          </gllink-stub>
        </div>
         
        <commit-stub
          commitref="[object Object]"
          commiturl="https://22878.qa-tunnel.gitlab.info/root/minimal-ruby-app/commit/63492726c2264a0277141d6a6573c3d22ecd7de3"
          shortsha="63492726"
          show-branch="true"
          showrefinfo="true"
          title="Improve code quality"
        />
      </div>
       
      <div
        class="col-sm-5 pl-0 text-right align-self-center d-none d-sm-block"
      >
        <timeago-stub
          time="2019-02-20T16:15:40.122Z"
          tooltiptext="Finished"
        />
         
        <alerts-stub
          count="1"
          lastalert="[object Object]"
        />
      </div>
       
      <!---->
    </div>
  </div>
</div>
`;
